### PR TITLE
engine: fix rocksdb iterator's behavior of valid

### PR DESCRIPTION
**Pull request by hicqu**
###  What have you changed?

According to rocksdb's document, errors should be handled correctly when an iterator becomes invalid. However this logic is missed in rust-rocksdb. This PR adds error handling for it.

###  What is the type of the changes?
Bug fix.

###  How is the PR tested?
Unit tests.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
No.

###  Does this PR affect `tidb-ansible`?
No.

###  Refer to a related PR or issue link (optional)
Please also take a look at https://github.com/tikv/rust-rocksdb/pull/411.

### CODE REVIEW

File: components/engine/src/iterable.rs [hunk 1]
```diff
@@ -128,32 +128,35 @@ pub trait Iterable {
     }
 
     // Seek the first key >= given key, if not found, return None.
+    // TODO: Make it zero-copy.
```
Comment (siddontang) [inline components/engine/src/iterable.rs]: how can we avoid this? @breeswish

File: components/engine/src/util.rs [hunk 2]
```diff
@@ -50,21 +50,17 @@ pub fn delete_all_in_range_cf(
...
                 // Can't use write_without_wal here.
                 // Otherwise it may cause dirty data when applying snapshot.
                 db.write(&wb)?;
                 wb.clear();
             }
-
-            if !it.next() {
-                break;
-            }
+            it_valid = it.next().unwrap();
```
Comment (Connor1996) [inline components/engine/src/util.rs]: How about 
```
loop {
    ...
    if !it.next()? { break; }
}
```

File: components/engine/src/iterable.rs (hunk 1 above)
Comment (Connor1996) [reply components/engine/src/iterable.rs]: seems we can just return `Result<Option<&[u8], &[u8]>>`

Comment (hicqu) [reply components/engine/src/iterable.rs]: Yes, we can do it later in an individual PR.

File: components/engine/src/iterable.rs [hunk 3]
```diff
@@ -128,32 +128,35 @@ pub trait Iterable {
...
-    it.seek(start_key.into());
-    while it.valid() {
-        let r = f(it.key(), it.value())?;
-
-        if !r || !it.next() {
-            break;
-        }
+    let mut it_valid = it.seek(start_key.into())?;
+    while it_valid {
+        it_valid = f(it.key(), it.value())? && it.next()?;
```
Comment (zhangjinpeng1987) [inline components/engine/src/iterable.rs]: it_valid means if iterator is valid, but here you miss use it.

File: components/engine_traits/src/iterable.rs [hunk 4]
```diff
@@ -11,42 +11,26 @@ pub enum SeekKey<'a> {
...
+    fn seek_to_last(&mut self) -> Result<bool> {
         self.seek(SeekKey::End)
     }
 
-    fn prev(&mut self) -> bool;
-    fn next(&mut self) -> bool;
+    fn prev(&mut self) -> Result<bool>;
+    fn next(&mut self) -> Result<bool>;
 
+    /// Must be called when `self.valid() == Ok(true)`.
```
Comment (zhangjinpeng1987) [inline components/engine_traits/src/iterable.rs]: s/Must/Only

File: components/engine_traits/src/iterable.rs [hunk 5]
```diff
@@ -119,35 +99,31 @@ where
...
-
-    fn next(&mut self) -> Option<Kv> {
-        let kv = self.0.kv();
-        if kv.is_some() {
-            I::next(&mut self.0);
-        }
-        kv
+    let mut it_valid = it.seek(SeekKey::Key(start_key))?;
+    while it_valid {
+        it_valid = f(it.key(), it.value())? && it.next()?;
```
Comment (zhangjinpeng1987) [inline components/engine_traits/src/iterable.rs]: Miss use of it_valid.

File: components/engine_traits/src/iterable.rs [hunk 6]
```diff
@@ -119,35 +100,31 @@ where
...
+pub fn collect<I: Iterator>(mut it: I) -> Vec<(Vec<u8>, Vec<u8>)> {
+    let mut v = Vec::new();
+    let mut it_valid = it.valid().unwrap();
+    while it_valid {
+        let kv = (it.key().to_vec(), it.value().to_vec());
+        v.push(kv);
+        it_valid = it.next().unwrap();
+    }
+    v
+}
```
Comment (brson) [inline components/engine_traits/src/iterable.rs]: Nice observation to remove the `as_std` adapter.

**Comment by brson** [comment]
Thanks for updating the engine_traits crate. Looks great.

**Pull request merged**